runtime/j9vm/jvm.c

```diff
@@ -4549,27 +4590,30 @@ JVM_RegisterSignal(jint sigNum, void* handler)
 		/* Don't allow user to register a native handler since
 		 * the signal is already used by the VM.
 		 */
-	} else if (J9_ARE_NO_BITS_SET(javaVM->sigFlags, J9_SIG_XRS)
-			&& isSignalSpecial(sigNum)) {
+		return (void *)J9_SIG_ERR;
+	} else if (isSpecialSignal && J9_ARE_ANY_BITS_SET(javaVM->sigFlags, J9_SIG_XRS | J9_SIG_XRS_ASYNC)) {
 		/* Don't allow user to register a native handler since
 		 * the signal is already used by the VM.
 		 */
+		return (void *)J9_SIG_ERR;
+	} else if (isSpecialSignal && isSignalIgnored(sigNum)) {
+		return (void *)J9_SIG_IGNORED;
 	} else {
 		/* Register the signal */
 #if defined(WIN32)
 		if ((void *)J9_PRE_DEFINED_HANDLER_CHECK == handler) {
-			handler = (void *)dummySignalHandler;
+			handler = (void *)defaultSignalHandler;
 		}
 		oldHandler = OMRSIG_SIGNAL(sigNum, handler);
 #else /* defined(WIN32) */
-		sigemptyset(&newSignalAction.sa_mask);
+		sigfillset(&newSignalAction.sa_mask);
 #if !defined(J9ZTPF)
-		newSignalAction.sa_flags = SA_RESTART;
+		newSignalAction.sa_flags = SA_RESTART | SA_SIGINFO;
 #else /* !defined(J9ZTPF) */
 		newSignalAction.sa_flags = 0;
 #endif /* !defined(J9ZTPF) */
 		if ((void *)J9_PRE_DEFINED_HANDLER_CHECK == handler) {
-			newSignalAction.sa_handler = dummySignalHandler;
+			newSignalAction.sa_handler = (void *)defaultSignalHandler;
 		} else {
 			newSignalAction.sa_handler = (void (*)(int))handler;
 		}
```

Comment:
I wonder if omrsig_set_async_signal_handler() can be used instead of OMRSIG_SIGACTION(), as this will take care of creating another thread.

Reply:
`omrsig_set_async_signal_handler` doesn't work with all the signals. it only works with these signals: `QUIT`, `ABRT`, `TERM`, `RECONFIG` and `XFSZ`.

Reply:
It could be fixed to handle all the shutdown signals.

Reply:
`JVM_RegisterSignal` needs to return the old handler while registering a new handler. `omrsig_set_async_signal_handler` will register a new handler to a signal but will not unregister the old handler. So, we would need to store the current handler for all signals, and unregister the current handler when a new handler is registered. There will also be an issue if a single handler is registered with multiple signals. Multiple calls to `omrsig_set_async_signal_handler` for the same handler will override the previous setting.